### PR TITLE
JSUI-2696 Changed SuggestionsManager's mergeSuggestion to use QueryProcessor

### PR DESCRIPTION
**Pull request by btaillon**
Changed SuggestionsManager's mergeSuggestion to use QueryProcessor instead and changed mergeSuggestion into an asynchronous function

https://coveord.atlassian.net/browse/JSUI-2696

[Deploy](https://dashboard.heroku.com/pipelines/a3535101-5bbf-4a5b-a909-47fcf8c9f149)

### CODE REVIEW

File: src/magicbox/MagicBox.ts [hunk 1]
```diff
@@ -191,10 +191,9 @@ export class MagicBoxInstance {
     };
   }
 
-  public showSuggestion() {
-    this.suggestionsManager.mergeSuggestions(this.getSuggestions != null ? this.getSuggestions() : [], suggestions => {
-      this.updateSuggestion(suggestions);
-    });
+  public async showSuggestion() {
+    const suggestions = await this.suggestionsManager.mergeSuggestions(this.getSuggestions != null ? this.getSuggestions() : []);
```
Comment (samisayegh) [inline src/magicbox/MagicBox.ts]: Why might `this.suggestions()` not be defined? `this.getSuggestions != null`

File: src/magicbox/MagicBox.ts [hunk 2]
```diff
@@ -235,11 +234,10 @@ export class MagicBoxInstance {
...
   }
 
-  public clearSuggestion() {
-    this.suggestionsManager.mergeSuggestions([], suggestions => {
-      this.updateSuggestion(suggestions);
-    });
+  public async clearSuggestion() {
     this.inputManager.setWordCompletion(null);
+    const suggestions = await this.suggestionsManager.mergeSuggestions([]);
+    this.updateSuggestion(suggestions);
```
Comment (samisayegh) [inline src/magicbox/MagicBox.ts]: I am curious what passing an empty array to `this.suggestionsManager.mergeSuggestions` does?

Also, are we storing suggestions inside of the `SuggestionsManager` instance as well as this MagicBox class? If so, it would better to keep one source of truth, likely inside the `SuggestionsManager`, to avoid bugs where we forget to update the local copy inside `MagicBox`.

File: src/magicbox/SuggestionsManager.ts [hunk 3]
```diff
@@ -141,73 +143,13 @@ export class SuggestionsManager {
...
-      })
-      .catch(() => {
-        return null;
-      });
+  public async mergeSuggestions(suggestions: Array<Promise<Suggestion[]> | Suggestion[]>) {
+    const { results, status } = await this.suggestionsProcessor.processQueries(suggestions);
+    if (status === ProcessingStatus.Overriden) {
+      return [];
+    }
+    this.updateSuggestions(results);
```
Comment (samisayegh) [inline src/magicbox/SuggestionsManager.ts]: Related to my comment above,
1- What does passing an empty array to this method do?
2- On line 151, we update the suggestions. Can we retrieve these values from the class by exposing a getter rather than returning the results and storing them in the MagicBox class? My feeling right now is this method should be `void`.

Comment (samisayegh) [reply src/magicbox/SuggestionsManager.ts]: Could it be easier for others working with the code to expose an `emptySuggestions` or `clearSuggestions` method?

```
this.suggestionsManager.clearSuggestions()
```

When I see `suggestionsManager.mergeSuggestions([])`, my first intuition is that the method will merge what I pass with the rest of the suggestions.

**Comment by ThibodeauJF** [comment]
Nice!

File: src/magicbox/SuggestionsManager.ts (hunk 3 above)
Comment (btaillon) [reply src/magicbox/SuggestionsManager.ts]: Good point, I'll rename the method and add clearSuggestions.

File: src/magicbox/MagicBox.ts [hunk 4]
```diff
@@ -191,10 +189,9 @@ export class MagicBoxInstance {
     };
   }
 
-  public showSuggestion() {
-    this.suggestionsManager.mergeSuggestions(this.getSuggestions != null ? this.getSuggestions() : [], suggestions => {
-      this.updateSuggestion(suggestions);
-    });
+  public async showSuggestion() {
+    await this.suggestionsManager.receiveSuggestions(this.getSuggestions());
```
Comment (samisayegh) [inline src/magicbox/MagicBox.ts]: What do you think of `addSuggesstions` or `pushSuggestions` or `processSuggestions` for the method name?

File: src/magicbox/MagicBox.ts [hunk 5]
```diff
@@ -191,10 +189,9 @@ export class MagicBoxInstance {
...
+    this.updateSuggestion(this.suggestionsManager.suggestions);
```
Comment (samisayegh) [inline src/magicbox/MagicBox.ts]: So do we ultimately we need to keep a local copy of the suggestions inside the magic box? I see on lines 250 and 269 that we are retrieving them from the suggestions manager?

File: src/magicbox/MagicBox.ts [hunk 6]
```diff
@@ -212,7 +209,6 @@ export class MagicBoxInstance {
   }
 
   private updateSuggestion(suggestions: Suggestion[]) {
-    this.lastSuggestions = suggestions;
     const firstSuggestion = this.getFirstSuggestionText();
     this.inputManager.setWordCompletion(firstSuggestion && firstSuggestion.text);
     this.onsuggestions && this.onsuggestions(suggestions);
```
Comment (samisayegh) [inline src/magicbox/MagicBox.ts]: Why might `onsuggestions` not be a function? Also, should the name be camel-cased `onSuggestions`?

File: src/magicbox/SuggestionsManager.ts [hunk 7]
```diff
@@ -28,11 +30,21 @@ export enum Direction {
   Right = 'Right'
 }
 
+type KeyboardFocus =
+  | {
+      type: 'suggestion';
+      suggestion: Suggestion;
```
Comment (samisayegh) [inline src/magicbox/SuggestionsManager.ts]: Rather than having two different property names holding two different types, could we use a common interface that both the `Suggestion` class and `ISearchResultPreview` implement/extend respectively?

Also, maybe the name of the type could be more specific/descriptive. e.g.

```
export interface IQuerySuggestPreviewFocusArgs = { // just a placeholder name, feel free to shorten/change
     property1: string;
     property2: string;
}

ISearchResultPreview extends IQuerySuggestPreviewFocusArgs {}
Suggestion implements IQuerySuggestPreviewFocusArgs
```

By reducing the type to one common interface, upstream code does not have to check the `type` property of the payload. This eliminates conditionals, creates one code path, and makes the code easier to test.

File: src/magicbox/MagicBox.ts (hunk 4 above)
Comment (btaillon) [reply src/magicbox/MagicBox.ts]: I like those a lot better. I'll replace `showSuggestion` with `addSuggestions`.

Also, this made me notice that the method `clear` calls `showSuggestion` instead of `clearSuggestions`, so I'll see if it's safe to replace that too.

File: src/magicbox/MagicBox.ts (hunk 5 above)
Comment (btaillon) [reply src/magicbox/MagicBox.ts]: I'm not exactly sure what you mean by that. Do you mean it would be preferable to keep them in MagicBox instead? Or that we should avoid MagicBox interacting with `Suggestion` altogether? If you mean the latter, it may involve moving `InputManager` in `SuggestionsManager`.
I won't change anything until I understand.

File: src/magicbox/MagicBox.ts (hunk 6 above)
Comment (btaillon) [reply src/magicbox/MagicBox.ts]: > Why might `onsuggestions` not be a function? 

`onsuggestions` is set at the same time that `MagicBox` gets instantiated, so you're right that it's always a function. I'll remove this check.

> Also, should the name be camel-cased `onSuggestions`?

You're right, I don't see why it couldn't be camel-cased. I'll change that.

File: src/magicbox/MagicBox.ts (hunk 5 above)
Comment (samisayegh) [reply src/magicbox/MagicBox.ts]: I see now on line 215 that we are no longer updating a local copy. In this case, is `updateSuggestions` still a good name for what the method is doing?

I see that we are:
1- Setting the input text to the suggestion text.
2- Calling `onsuggestions`. Not sure what this method is doing?
3- Adding an `onSelect` handler on every suggestion where the handler is not defined and the suggestion contains text.

That's not being conveyed by the name right now.

File: src/magicbox/SuggestionsManager.ts (hunk 7 above)
Comment (btaillon) [reply src/magicbox/SuggestionsManager.ts]: I like this, however, `Suggestion` has a property named `dom` and `ISearchResultPreview`'s equivalent property is named `element`. `element` is technically a better name but changing `dom` would affect a lot of references. Which would be better?

File: src/magicbox/MagicBox.ts (hunk 5 above)
Comment (btaillon) [reply src/magicbox/MagicBox.ts]: Good point, I'll rename `updateSuggestion` to `addSelectEventHandlers` and I'll move some code out of this method.

File: src/magicbox/SuggestionsManager.ts (hunk 7 above)
Comment (samisayegh) [reply src/magicbox/SuggestionsManager.ts]: One option could add a new public getter to one of the objects that returns `this.dom`. e.g.

```
// Suggestion.ts

public get element() {
    return this.dom;
}
```

Comment (btaillon) [reply src/magicbox/SuggestionsManager.ts]: Unfortunately, `Suggestion` is an interface and transforming it into a class would also involve several changes in other files. Would making a wrapping class for `SuggestionsManager` be a good idea?

E.g.:
```ts
export class ActiveSuggestion implements IQuerySuggestPreviewFocusArgs {
    public get element() { ... }
    constructor(private suggestion: Suggestion) { ... }
}
```

File: src/magicbox/MagicBox.ts [hunk 8]
```diff
@@ -235,24 +231,27 @@ export class MagicBoxInstance {
...
+      this.inputManager.setResult(this.displayedResult, this.getFirstSuggestionText());
     } else {
       this.inputManager.setResult(this.grammar.parse(suggestion.text).clean(), suggestion.text);
     }
   }
 
-  private getFirstSuggestionText(): Suggestion {
-    return find(this.lastSuggestions, suggestion => suggestion.text != null);
+  private getFirstSuggestionWithText() {
+    return find(this.suggestionsManager.suggestions, suggestion => suggestion.text != null);
```
Comment (ThibodeauJF) [inline src/magicbox/MagicBox.ts]: Not a fan of the `!= null` here. Could maybe just be `!!suggestion.text`

File: src/magicbox/SuggestionsManager.ts [hunk 9]
```diff
@@ -350,17 +316,18 @@ export class SuggestionsManager {
   }
 
   private async moveWithinSuggestion(direction: Direction) {
-    const currentlySelected = $$(this.element).find(`.${this.options.selectedClass}`);
-    const selectables = $$(this.element).findAll(`.${this.options.suggestionClass}`);
-    const currentIndex = indexOf(selectables, currentlySelected);
+    const currentlySelected =
+      this.keyboardFocusedElement && this.keyboardFocusedElement.type === 'suggestion' ? this.keyboardFocusedElement.suggestion : null;
+    const selectables = this.displayedSuggestions;
+    const currentIndex = currentlySelected !== null ? indexOf(selectables, currentlySelected) : -1;
```
Comment (ThibodeauJF) [inline src/magicbox/SuggestionsManager.ts]: `currentlySelected !== null` -> `currentlySelected`

File: src/magicbox/SuggestionsManager.ts [hunk 10]
```diff
@@ -439,8 +389,7 @@ export class SuggestionsManager {
     input.setAttribute('aria-autocomplete', 'list');
   }
 
-  private htmlElementIsSuggestion(selected: HTMLElement) {
-    const omniboxSelectables = $$(this.element).findAll(`.${this.options.suggestionClass}`);
-    return indexOf(omniboxSelectables, selected) > -1;
+  private getSuggestionFromElement(element: HTMLElement) {
+    return find(this.displayedSuggestions, suggestion => suggestion.dom === element);
```
Comment (ThibodeauJF) [inline src/magicbox/SuggestionsManager.ts]: I find it a bit confusing that `dom` is of type HTMLElement when the Dom type is also commonly used in many places

File: src/magicbox/MagicBox.ts (hunk 8 above)
Comment (btaillon) [reply src/magicbox/MagicBox.ts]: Good point, fixing that.

File: src/magicbox/SuggestionsManager.ts (hunk 9 above)
Comment (btaillon) [reply src/magicbox/SuggestionsManager.ts]: Good point, fixing that one too.

File: src/magicbox/SuggestionsManager.ts (hunk 10 above)
Comment (btaillon) [reply src/magicbox/SuggestionsManager.ts]: I agree. Should it be renamed in this PR or be part of [JSUI-2700](https://coveord.atlassian.net/browse/JSUI-2700)?

File: src/magicbox/SuggestionsManager.ts (hunk 3 above)
Comment (btaillon) [reply src/magicbox/SuggestionsManager.ts]: UPDATE: this fix was moved to [JSUI-2700](https://coveord.atlassian.net/browse/JSUI-2700).

File: src/magicbox/MagicBox.ts [hunk 11]
```diff
@@ -235,24 +232,27 @@ export class MagicBoxInstance {
...
-      this.inputManager.setResult(this.displayedResult, suggestion && suggestion.text);
+      this.inputManager.setResult(this.displayedResult, this.getFirstSuggestionText());
     } else {
       this.inputManager.setResult(this.grammar.parse(suggestion.text).clean(), suggestion.text);
     }
   }
 
-  private getFirstSuggestionText(): Suggestion {
-    return find(this.lastSuggestions, suggestion => suggestion.text != null);
+  private getFirstSuggestionWithText(): Suggestion {
```
Comment (samisayegh) [inline src/magicbox/MagicBox.ts]: I would convert this method and the next one to getters. e.g.
```
private get firstSuggestionWithText(): Suggestion {}
```

File: src/magicbox/SuggestionsManager.ts [hunk 12]
```diff
@@ -128,7 +130,7 @@ export class SuggestionsManager {
 
   public selectAndReturnKeyboardFocusedElement(): HTMLElement {
     const selected = this.keyboardFocusedElement;
-    if (selected != null) {
+    if (!!selected) {
```
Comment (samisayegh) [inline src/magicbox/SuggestionsManager.ts]: I'm seeing the double `!!` here and on line 250 in the previous file. Any reason not to just do:
```
if (selected) {
    ...
}
```
and

```
return find(this.lastSuggestions, suggestion => suggestion.text);
```

File: src/magicbox/SuggestionsManager.ts [hunk 13]
```diff
@@ -141,73 +143,17 @@ export class SuggestionsManager {
...
-        stillNeedToResolve = false;
-      };
-
-      if (suggestions.length == 0) {
-        onResolve();
-      }
-      if (suggestions == undefined) {
-        onResolve();
-      }
+  public async receiveSuggestions(suggestions: Array<Promise<Suggestion[]> | Suggestion[]>) {
```
Comment (samisayegh) [inline src/magicbox/SuggestionsManager.ts]: For array types, we usually use the bracket syntax. e.g. `Promise<Suggestion[]>[]`

Also, if Suggestion is an interface, we usually add an `I` in front. e.g. `ISuggestion`. This refactor could be in a separate PR though.

File: src/magicbox/SuggestionsManager.ts (hunk 12 above)
Comment (ThibodeauJF) [reply src/magicbox/SuggestionsManager.ts]: Yes, `!!` is very useful for casting to a boolean, but a `if` will already evaluate the condition to see it's "truthy" (here selected). So it will check if it's not `null`, `undefined`, `0`, etc.

File: src/magicbox/MagicBox.ts (hunk 11 above)
Comment (btaillon) [reply src/magicbox/MagicBox.ts]: Good idea, I'm converting them.

File: src/magicbox/SuggestionsManager.ts (hunk 12 above)
Comment (btaillon) [reply src/magicbox/SuggestionsManager.ts]: Woops, I just went by fast converting them all without paying much attention.

It's the kind of thing that should've already been a habit for me by now, given that before Javascript I programmed in Lua, which evaluated the exact same types as `true`...

### Tl;dr: Fixing it.

File: src/magicbox/SuggestionsManager.ts (hunk 13 above)
Comment (btaillon) [reply src/magicbox/SuggestionsManager.ts]: Good point, converting it from `Array<Promise<Suggestion[]> | Suggestion[]>` to `(Promise<Suggestion[]> | Suggestion[])[]`. I kinda wish Typescript also had symbol for promises.

As for `ISuggestion`, I was thinking of converting it in [JSUI-2700](https://coveord.atlassian.net/browse/JSUI-2700).
